doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,222 @@
+Especially for `libelektra-core` (the only place where `Key * errorKey` truly cannot be used), all functions must be side effect free.
+Therefore, the only possible cause of errors are invalid arguments.
+
+If we now assume the function is written such that it fails before it changes any argument's data, then we can infer that only a single error indicating return value is needed.
+Because the arguments are unchanged and the only possible cause of errors, we can distinguish between errors after the fact.
+
+As stated, many functions' only error cases are precondition violations.
+If the preconditions are simple enough, they can be checked beforehand if it is not clear whether the precondition holds.
+
+Take for example some hypothetical `keyValue` function, in a world where we have a `key->value` container for COW that should always be present.
+Looking just at the error checks, without even considering the function signature:
+
+```c
+if (key == NULL) {
+// error 1: precondition violation
+}
+
+if (key->value == NULL) {
+// error 2: invalid Key, COW container missing
+}
+
+// success: return value
+return key->value->data;
+```
+
+If we want separate error indications for both error cases and a nice API we immediately hit a problem.
+The obvious return type is `const void *`, because that's the type of the key value (and we want it const).
+But that's a pointer, so only `NULL` could be used for errors.
+The issue is, `NULL` may be a valid key value.
+So what do we do?
+
+Essentially, we do nothing, "everything is fine":
+
+```c
+const void * keyValue (Key * key) {
+  if (key == NULL) {
+    return NULL;
+  }
+
+  assert(key->value != NULL);
+  if (key->value == NULL) {
+    return NULL;
+  }
+
+  return key->value->data;
+}
+```
+
+We return `NULL` for both error cases and even let the errors overlap with a legitimate success value.
+In the first case, it's fine for multiple reasons:
+
+- Many callers won't care, if the value inside the key is `NULL` or the key itself is `NULL`.
+  For example, `keyValue (ksLookup (ks, name)) == NULL` is probably fine either way.
+- If the caller does care, they can easily check `if (key == NULL)` before _or_ after the `keyValue` call.
+
+The second case has a different reason.
+The error is not the caller's fault.
+It is an internal issue the caller can do nothing about, `key->value` should always be there.
+That's why we add an `assert` before the check.
+This will trigger in a debug build, where the internal error can be fixed.
+For release builds, the `assert` won't be there.
+Instead, we return `NULL`, so the caller can do something and doesn't have to deal with the internal bug.
+
+This overlapping of error and success values has especially positive effects, when we talk about boolean functions:
+
+```c
+// returns 1 if key is "foo", 0 if key is not "foo", and -1 if key == NULL
+int keyIsFooInt (Key * key);
+
+// returns true if key is "foo", false otherwise (including if key == NULL)
```

Comment:
Seems to me quite unrelated.

Reply:
The connection to this decision is: Not all functions need a specific error value. Sometimes it is good enough or maybe even better, to use a default value for things like `NULL` arguments.